src/solver.rs

```diff
@@ -166,17 +172,29 @@ pub fn resolve<P: Package, V: Version>(
                 "Root package depends on itself at a different version?".into(),
             ));
         }
-        state.partial_solution.add_version(p, v, &dep_incompats);
+        state
+            .partial_solution
+            .add_version(next.clone(), v, &dep_incompats);
     }
 }
 
 /// Trait that allows the algorithm to retrieve available packages and their dependencies.
 /// An implementor needs to be supplied to the [resolve] function.
 pub trait DependencyProvider<P: Package, V: Version> {
-    /// Lists available versions for a given package.
-    /// The strategy of which version should be preferably picked in the list of available versions
-    /// is implied by the order of the list: first version in the list will be tried first.
-    fn list_available_versions(&self, package: &P) -> Result<Vec<V>, Box<dyn Error>>;
+    /// Chooses which package the [resolve] should decide on next.
+    /// If there are available versions for that package returns `Some(V)` to try.
+    /// If there are not available versions for that package returns `None`.
+    ///
+    /// This method holds a lot of control over the performance and behaviour of the algorithm.
+    /// It is normal to pick the package that has the fewest versions contained in the constraints,
+    /// as they are likely to find problems faster. Also normal is to pick the most recent matching
+    /// version.
+    ///
+    /// Note: the type `T` ensures that this returns an item from the `packages` argument.
+    fn make_decision<T: Borrow<P>, U: Borrow<Range<V>>>(
+        &self,
+        packages: impl Iterator<Item = (T, U)>,
+    ) -> Result<(T, Option<V>), Box<dyn Error>>;
```

Comment:
I propose to provide a default implementation for this method with the heuristic we were using previously.

This way we still give control to the users who want it, but provide the option to leave the way we pick packages as the library suggests.

Reply:
Do you want to keep the `list_available_versions`? Because without that I don't know how to build a default implementation. It was removed as we otherwise don't need it.

Reply:
I don't think we should keep `list_available_versions` in the interface. Maybe we could have a free function (or collection of free functions) that could help. Something like `pick_first_strategy`, `pick_lowest_number_of_versions_strategy` etc. The names are terrible, but that's just for the sake of the example

Reply:
I was sceptical, but gave it a try. And it worked out really well! Take a look and let me know what you think.

Reply:
> I don't think we should keep list_available_versions in the interface.

Why not? What if our interface is 3 functions:
1) `make_decision()`
2) `list_available_versions()`
3) `get_dependencies()`

The first one has default implementation in terms on `list_available_versions`, no need to change it right away for the algorithm to work but the ability to customize it is there.

The idea is somewhat similar to what iterators to with the `next()` function. You define it and get some others for free. Here we would implement 1) in terms of 2).

The only not so pretty thing about that is if the user comes up with a way to redefine `make_decision` without using `list_available_versions`, but the API still requires it. In that case they could leave it `unimplemented!`. Is it an API smell? Hmmm.

Reply:
I forgot about `should_cancel` but the point stands.

Reply:
It is a workable solution, if we had a lot of users of the current api that would be a wonderful way to add an opt in for performance without breaking existing users. The boilerplate required to use the helper function, did not end up being so onerous IMO. I think it is a code smell to require our users to provide something that we will not be using.
I don't think "avoid the breaking change" justifies that smell  given our user base, nor do I think "make it easy to get started" justifies it with the helper function. But it can work if that is what we want to do.

Reply:
> I don't think "avoid the breaking change" justifies that smell given our user base

My motivation is not avoiding breaking changes, but rather the easiness of API usage.

Before this PR it was crystal clear what the algorithm needed from the user to function (from the user's perspective):
1) A way to get dependencies for a given package version.
2) A way to get all versions for the package.

All makes sense why it is needed and how to implement it.

Now:
1) Unchanged.
2) Hm, actually you need to implement part of our algorithm to use our library.

How would people do that? What if they don't want to understand how exactly we do the decision making?

The only practical solution is to copy paste the code from `OfflineDependencyProvider`.
I'm afraid that requiring every user to do that is too much of an ergonomic hit.

Does anyone else share my concerns?

Reply:
I get your point about ease of use. However, I think people not wanting to use `OfflineDependencyProvider` and willing to define their own dependency provider are willing to read the implementation of an 11-lines function doing nothing more than a function call to a helper function. Also there is an even simpler [6-lines example implementation in the guide](https://pubgrub-rs-guide.netlify.app/pubgrub_crate/custom_dep_provider.html) on the section regarding custom implementation of a dependency provider.

If we get more users, and some of them point `make_decision` as being difficult to implement, I think I'd rather have a two different traits with different ease/control compromises (a bit like the [sandbox/element/document/application variants of an elm app](https://package.elm-lang.org/packages/elm/browser/latest/)) than having a single trait with an ambiguous API.

Reply:
> If we get more users, and some of them point make_decision as being difficult to implement, I think I'd rather have a two different traits with different ease/control compromises (a bit like the sandbox/element/document/application variants of an elm app) than having a single trait with an ambiguous API.

Actually that's what my girlfriend suggested as well! :laughing: I was pretty impressed because she's sophomore (2nd year in the uni) and is doing that as a way to change her profession, meaning she has very little experience in programming yet.

In any case, doing that is out of scope of this PR and needs a separate discussion; I'm totally okay merging this version!

Reply:
I'm glad we had this discussion as this is an interesting design problem I haven't had to think of before, thank you @Eh2406 and @mpizenberg.

Reply:
you have a wise girlfriend @aleksator  ahah, keep her! XD